### PR TITLE
Update mypy version to 1.0.1

### PR DESCRIPTION
**Pull request by koxudaxi**
The PR updates mypy version to 1.0.1 for performance.

## Related Issues
 https://github.com/pydantic/pydantic/issues/5083

## Motivation
The current CI test is a little slow for mypy tests.
Mypy 1.0.0 is improved performance.
from the blog 👇

https://mypy-lang.blogspot.com/2023/02/mypy-10-released.html
_Performance Improvements
Mypy 1.0 is up to 40% faster than mypy 0.991 when type checking the Dropbox internal codebase. We also set up a daily job to measure the performance of the most recent development version of mypy to make it easier to track changes in performance._

## Easy benchmarks
`coverage run -m pytest --durations=10 tests/mypy`

```bash
tail -n3 mypy*.log 
==> mypy0.961.log <==
Results (136.28s):
        38 passed
         5 xfailed

==> mypy0.982.log <==
Results (286.70s):
        38 passed
         5 xfailed

==> mypy1.0.0.log <==
Results (75.57s):
        38 passed
         5 xfailed

==> mypy1.0.1.log <==
Results (75.38s):
        38 passed
         5 failed
```

### CODE REVIEW

File: pydantic/_internal/_validators.py [hunk 1]
```diff
@@ -157,7 +157,9 @@ def __call__(self, __input_value: int | float | str, **_kwargs: Any) -> Decimal:
                 raise PydanticKnownError('finite_number')
 
             if self.check_digits:
-                if exponent >= 0:
+                if isinstance(exponent, str):
+                    raise PydanticKnownError('finite_number')
+                elif exponent >= 0:
```
Comment (koxudaxi) [inline pydantic/_internal/_validators.py:162]: mypy show  the type error of comparing `literal` and  `int`
`check_digits` feature doesn't work for Inf and NaN values.

File: pydantic/json.py [hunk 2]
```diff
@@ -35,7 +35,8 @@ def decimal_encoder(dec_value: Decimal) -> Union[int, float]:
     >>> decimal_encoder(Decimal("1"))
     1
     """
-    if dec_value.as_tuple().exponent >= 0:
+    exponent = dec_value.as_tuple().exponent
+    if isinstance(exponent, int) and exponent >= 0:
         return int(dec_value)
```
Comment (koxudaxi) [inline pydantic/json.py:40]: I can change `isinstance` to use try/exception for performance.

**Comment by koxudaxi** [comment]
I added some conditions when the Decimal has `Inf` and `NaN`.
If we should separate the PR for the change, I will create it.

File: pydantic/_internal/_validators.py (hunk 1 above)
Comment (samuelcolvin) [reply pydantic/_internal/_validators.py:162]: i think this is fine, but please add a comment explaining why we need the above.

File: pydantic/json.py (hunk 2 above)
Comment (samuelcolvin) [reply pydantic/json.py:40]: I think this is fine, we'll need to change this code anyway.

**Review by samuelcolvin — APPROVED** [review]
otherwise LGTM.

This is great, thank you.

**Pull request merged**